packages/backend/src/open_payments/auth/grant.ts

```diff
@@ -108,3 +142,58 @@ export class Grant {
     }
   }
 }
+
+// Export for testing
+export function getInterval(
+  repeatingInterval: string,
+  target: Date
+): Interval | undefined {
+  const parts = repeatingInterval.split('/')
+  assert.ok(parts.length === 3)
+
+  let repetitions: number | undefined
+  if (parts[0].length > 1 && parts[0][1] !== '-') {
+    repetitions = Number(parts[0].slice(1))
+  } else if (
+    (parts[0].length === 1 && parts[0][0] === 'R') ||
+    (parts[0].length > 1 && parts[0][1] === '-')
+  ) {
+    repetitions = Infinity
+  }
+  if (repetitions === undefined || isNaN(repetitions)) return
+
+  const interval = Interval.fromISO(`${parts[1]}/${parts[2]}`)
+  if (!interval.start) return
+  if (interval.contains(DateTime.fromJSDate(target))) return interval
+
+  let duration: string
+  let forward: boolean
+  if (parts[1].length > 1 && parts[1][0] === 'P') {
+    duration = parts[1]
+    forward = false
+  } else if (parts[2].length > 1 && parts[2][0] === 'P') {
+    duration = parts[2]
+    forward = true
+  } else {
+    duration = interval.toDuration().toString()
+    forward = true
+  }
+
+  const intervals = [interval]
```

Comment:
Could `interval` just be reassigned in the loop instead of building an `intervals` array?